app/src/main/java/com/eveningoutpost/dexdrip/NSEmulatorReceiver.java

```diff
@@ -247,83 +248,116 @@ private JSONObject extractParams(final Bundle bundle) {
+                }
+                computed_bg = json_object.getJSONArray(Intents.HISTORIC_BG);
+                history_bg_vals = new int[computed_bg.length()];
+                for (int i = 0; i < computed_bg.length(); i++) {
+                    history_bg_vals[i] = computed_bg.getInt(i);
+                }
+            }
         } catch (JSONException e) {
             Log.e(TAG, "Error JSONException ", e);
             return;
         }
-        if(decoded_buffer == null) {
+        if (decoded_buffer == null) {
             Log.e(TAG, "Error could not get decoded_buffer");
             return;
         }
-        
+
         // Does this throws exception???
         byte[] fram_data = Base64.decode(decoded_buffer, Base64.NO_WRAP);
-        byte []patchUid = Base64.decode(patchUidString, Base64.NO_WRAP);
-        byte []patchInfo = Base64.decode(patchInfoString, Base64.NO_WRAP);
-        LibreOOPAlgorithm.handleOop2DecodeFramResult(tagId, CaptureDateTime, fram_data, patchUid, patchInfo);
+        byte[] patchUid = Base64.decode(patchUidString, Base64.NO_WRAP);
+        byte[] patchInfo = Base64.decode(patchInfoString, Base64.NO_WRAP);
+        LibreOOPAlgorithm.handleOop2DecodeFramResult(tagId, CaptureDateTime, fram_data, patchUid, patchInfo, trend_bg_vals, history_bg_vals);
     }
-    
+
     private void handleOop2DecodeBleResult(Bundle bundle) {
-        if(Pref.getBooleanDefaultFalse("external_blukon_algorithm")) {
+        if (Pref.getBooleanDefaultFalse("external_blukon_algorithm")) {
             Log.e(TAG, "External OOP algorithm is on, ignoring ble decrypted data.");
             return;
         }
         JSONObject json_object = extractParams(bundle);
-        if(json_object == null) {
+        if (json_object == null) {
             return;
         }
         String decoded_buffer;
         String patchUidString;
+        JSONArray computed_bg;
         long CaptureDateTime;
+
+        int[] trend_bg_vals = null;
+        int[] history_bg_vals = null;
+
         try {
             decoded_buffer = json_object.getString(Intents.DECODED_BUFFER);
             patchUidString = json_object.getString(Intents.PATCH_UID);
             CaptureDateTime = json_object.getLong(Intents.LIBRE_DATA_TIMESTAMP);
+            if (json_object.has(Intents.TREND_BG) && json_object.has(Intents.HISTORIC_BG)) {
+                computed_bg = json_object.getJSONArray(Intents.TREND_BG);
+                trend_bg_vals = new int[computed_bg.length()];
+                for (int i = 0; i < computed_bg.length(); i++) {
+                    trend_bg_vals[i] = computed_bg.getInt(i);
+                }
+                computed_bg = json_object.getJSONArray(Intents.HISTORIC_BG);
+                history_bg_vals = new int[computed_bg.length()];
+                for (int i = 0; i < computed_bg.length(); i++) {
+                    history_bg_vals[i] = computed_bg.getInt(i);
+                }
+            }
+
```

Comment:
It can be factored to one functoin, but it will have a boolean saying weather this is FRAM or BLE and then it will have 3-4 if branches testing what option is involved. I think that seperating the two pathes makes the code easier to read as each is handeling its own intent.

Reply:
Sorry, I cannot follow you. I've harmonized the two functions and a diff looks like:
![image](https://user-images.githubusercontent.com/1370732/153309715-e69eee03-7db5-4969-88ec-3e00ac677331.png)

As you can see, the inner part is completely identical. Extracting this into a private sub function increases the readability and maintainability quite a lot, IMHO.
[handleOop2DecodeFramResult.txt](https://github.com/NightscoutFoundation/xDrip/files/8036896/handleOop2DecodeFramResult.txt)
[handleOop2BluetoothEnableResult.txt](https://github.com/NightscoutFoundation/xDrip/files/8036897/handleOop2BluetoothEnableResult.txt)

Reply:
After thinking about a little more, these two functions can be merged into a single one and testing for `json_object.has(Intents.TAG_ID)` can distinguish the different handling.

Reply:
I do not agree that using one function for both is a good idea.
Right now there are two code flows: one for using ble data and one for using fram data.
Both are sent to OOP2 using different functions.
The code that handles them is in a function called handleOop2DecodeFramResult (for example). it starts with nsemulatorreciver where the intent is being handled, and continues to oop2 where data is handled.
There is a similar code for BLE.
Everyone reading the code can see that there is a path for BLE and a path for FRAM and it is easy to follow.
Now, if I would create one function from both of them 3 things will happen:

1. The code will not be clear. For example the name of the function will not tell if it is BT or FRAM. 
2. Drawing a code graph will show that the BT data might be called from a FRAM intent.  and vice versa.
3. Since one function has to handle an extra parameter, there will be ifs in the code to say if the parameter should exist or should not exist. The code becomes harder to be understood.

One more point to understand is that in no case we can use json_object.has(Intents.TAG_ID) to say if this should be a BLE data or FRAM data. Doing so, will cause the code to be very fragile. We will be using some unrelated parameter to decide if this is BT or FRAM. For example, In the case of a change in OOP2 we might have the code treat BLE data as FRAM data or vice versa. We can pass a boolean to make the decision more clear, but this will look ugly.

Bottom line, I don't file that removing 30 lines of code is worth the effort.